lib/engine.rb

```diff
@@ -1,64 +1,130 @@
 class Game
   attr_reader :state
 
   def initialize
     @state = true
+    @counter = 0
+  end
+
+  def increase_counter
+    @counter += 1
   end
 
-  # state to finish games
   def finish_game
     @state = false
   end
+
+  def line_checker(board, winner)
+    return unless @counter >= 4
+
+    vertical = vertical(board)
+
+    horizontal = horizontal(board)
+
+    diagonal = diagonal(board)
+
+    if vertical || horizontal || diagonal
+      finish_game
+      puts "Player #{winner} won!"
+    elsif @counter == 8
+      puts 'Draw!'
+    end
+  end
+
+  def vertical(board)
+    board[0][0].state == board[1][0].state && board[1][0].state == board[2][0].state ||
+      board[0][1].state == board[1][1].state && board[1][1].state == board[2][1].state ||
+      board[0][2].state == board[1][2].state && board[1][2].state == board[2][2].state
+  end
+
+  def horizontal(board)
+    board[0][0].state == board[0][1].state && board[0][1].state == board[0][2].state ||
+      board[1][0].state == board[1][1].state && board[1][1].state == board[1][2].state ||
+      board[2][0].state == board[2][1].state && board[2][1].state == board[2][2].state
+  end
+
+  def diagonal(board)
+    board[0][0].state == board[1][1].state && board[1][1].state == board[2][2].state ||
+      board[0][2].state == board[1][1].state && board[1][1].state == board[2][0].state
+  end
 end
 
 class Cell
-  attr_reader :state
-
   def initialize
     @state = nil
   end
 
-  # Method to mark X on board
+  attr_reader :state
+
+  attr_writer :state
+
   def make_cross
     @state = 'X'
   end
 
-  # Method to mark O on board
   def make_o
     @state = 'O'
   end
 end
 
-# Tic Tack Toe grid
 class Board
+  attr_reader :board_status
+
   def initialize
-    @board_status = {
-      1 => Cell.new,
-      2 => Cell.new,
-      3 => Cell.new,
-      4 => Cell.new,
-      5 => Cell.new,
-      6 => Cell.new,
-      7 => Cell.new,
-      8 => Cell.new,
-      9 => Cell.new
-    }
+    @board_status = Array.new(3) { Array.new(3) { Cell.new } }
+    @board_status.each_with_index do |arr, f_index|
+      arr.each_with_index do |elem, s_index|
+        print elem.state
+        elem.state = (f_index * 3 + s_index + 1).to_s
+      end
+    end
   end
 
   def change_cell_o(cell)
-    @board_status[cell].make_o
+    @board_status[cell[0]][cell[1]].make_o
   end
 
   def change_cell_cross(cell)
-    @board_status[cell].make_cross
+    @board_status[cell[0]][cell[1]].make_cross
+  end
+
+  def cell_empty?(cell)
+    @board_status[cell[0]][cell[1]].state == 'X' && @board_status[cell[0]][cell[1]].state == 'O' ? false : true
   end
 
   def print_board
-    (1..@board_status.length).each do |i|
-      print "#{@board_status[i].state.nil? ? i : @board_status[i].state} #{(i % 3).zero? && !i.zero? ? "\n" : ''}"
+    @board_status.each do |arr|
+      arr.each do |elem|
+        print elem.state
+        print ' '
+      end
+      puts "\n"
     end
   end
 end
+
+class InputManipulator
```

Comment:
[OPTIONAL]
Maybe try putting this class in a separate file as it is better separation of code.
If you really want to have it in this engine file you can always require it